Error when running mdbook build: Unexpected end of input while parsing minor version number

```
D:\book>mdbook.exe build
2024-03-28 21:36:37 [INFO] (mdbook::book): Book building has started
2024-03-28 21:36:37 [INFO] (mdbook::book): Running the pandoc backend
2024-03-28 21:36:37 [INFO] (mdbook::renderer): Invoking the "pandoc" renderer
thread 'main' panicked at src\pandoc.rs:69:56:
called `Result::unwrap()` on an `Err` value: Error("unexpected end of input while parsing minor version number")
stack backtrace:
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
2024-03-28 21:36:37 [ERROR] (mdbook::renderer): Renderer exited with non-zero return code.
2024-03-28 21:36:37 [ERROR] (mdbook::utils): Error: Rendering failed
2024-03-28 21:36:37 [ERROR] (mdbook::utils):    Caused By: The "pandoc" renderer failed
```

I have tried the following steps without success:

Initializing a book. `mdbook init book`
Confirming pandoc installation and accessibility in the command line.
Checking compatibility between mdbook and pandoc versions.
Updating mdbook and pandoc to the latest versions.
Verifying `book.toml` settings.

This time, another error occurs to me, again. just like him [](https://github.com/max-heller/mdbook-pandoc/issues/81)
```
2024-03-29 08:59:27 [INFO] (mdbook::book): Book building has started
2024-03-29 08:59:27 [INFO] (mdbook::book): Running the pandoc backend
2024-03-29 08:59:27 [INFO] (mdbook::renderer): Invoking the "pandoc" renderer
pandoc: book\pdf\src\intro.md: openBinaryFile: invalid argument (Invalid argument)
pandoc exited unsuccessfully
2024-03-29 08:59:27 [ERROR] (mdbook::renderer): Renderer exited with non-zero return code.
2024-03-29 08:59:27 [ERROR] (mdbook::utils): Error: Rendering failed
2024-03-29 08:59:27 [ERROR] (mdbook::utils):    Caused By: The "pandoc" renderer failed
```

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -108,7 +108,7 @@ impl mdbook::Renderer for Renderer {
             let ctx = pandoc::RenderContext {
                 book: &book,
                 mdbook_cfg: &ctx.config,
-                pandoc: pandoc::Context::new(pandoc_version.clone()),
+                pandoc: pandoc::Context::new(pandoc_version),
                 destination: book.destination.join(name),
                 output: profile.output_format(),
                 columns: profile.columns,

--- a/src/pandoc.rs
+++ b/src/pandoc.rs
@@ -1,7 +1,6 @@
-use std::{collections::BTreeMap, process::Command};
+use std::{collections::BTreeMap, fmt, num::ParseIntError, process::Command, str::FromStr};
 
 use anyhow::{anyhow, Context as _};
-use once_cell::sync::Lazy;
 
 pub mod extension;
 pub use extension::Extension;
@@ -12,18 +11,48 @@ pub use profile::Profile;
 mod renderer;
 pub use renderer::{Context as RenderContext, OutputFormat, Renderer};
 
-/// Defines compatible versions of Pandoc
-pub static VERSION_REQ: Lazy<semver::VersionReq> =
+/// Minimum compatible version of Pandoc
+const MINIMUM_VERSION: Version =
     // commonmark input format introduced in 1.14
-    Lazy::new(|| semver::VersionReq::parse(">=1.14").unwrap());
+    Version {
+        major: 1,
+        minor: 14,
+        patch: 0,
+    };
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Version {
+    major: u64,
+    minor: u64,
+    patch: u64,
+}
+
+impl fmt::Display for Version {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}.{}.{}", self.major, self.minor, self.patch)
+    }
+}
+
+impl FromStr for Version {
+    type Err = ParseIntError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut components = s.split('.').map(|component| component.parse());
+        let mut next_component = || components.next().unwrap_or(Ok(0));
+        Ok(Self {
+            major: next_component()?,
+            minor: next_component()?,
+            patch: next_component()?,
+        })
+    }
+}
 
 pub struct Context {
-    pub version: semver::Version,
+    pub version: Version,
     enabled_extensions: BTreeMap<Extension, extension::Availability>,
 }
 
 impl Context {
-    pub fn new(version: semver::Version) -> Self {
+    pub fn new(version: Version) -> Self {
         let mut this = Self {
             enabled_extensions: Default::default(),
             version,
@@ -47,7 +76,7 @@ impl Context {
     }
 }
 
-pub fn check_compatibility() -> anyhow::Result<semver::Version> {
+pub fn check_compatibility() -> anyhow::Result<Version> {
     let version = {
         let output = Command::new("pandoc")
             .arg("-v")
@@ -60,23 +89,19 @@ pub fn check_compatibility() -> anyhow::Result<semver::Version> {
         );
         let output = String::from_utf8(output.stdout).context("`pandoc -v` output is not UTF8")?;
         match output.lines().next().and_then(|line| line.split_once(' ')) {
-            Some(("pandoc", mut version)) => {
-                // Pandoc versions can contain more than three components (e.g. a.b.c.d).
-                // If this is the case, only consider the first three.
-                if let Some((idx, _)) = version.match_indices('.').nth(2) {
-                    version = &version[..idx];
-                }
-                semver::Version::parse(version.trim()).unwrap()
-            }
+            Some((_, version)) => version
+                .trim()
+                .parse()
+                .with_context(|| format!("failed to parse Pandoc version '{version}'"))?,
             _ => anyhow::bail!("`pandoc -v` output does not contain `pandoc VERSION`"),
         }
     };
-    if VERSION_REQ.matches(&version) {
+    if version >= MINIMUM_VERSION {
         Ok(version)
     } else {
         Err(anyhow!(
-            "mdbook-pandoc is incompatible with detected Pandoc version (requires version {}, but using {})",
-            *VERSION_REQ, version,
+            "mdbook-pandoc is incompatible with detected Pandoc version \
+            (requires at least {MINIMUM_VERSION}, but using {version})"
         ))
     }
 }
@@ -87,6 +112,8 @@ mod tests {
         io::Write,
         process::{Command, Stdio},
     };
+
+    use super::*;
 
     // Canary to detect if Pandoc ever adds native support for lists
     // nested more than four layers deep when rendering to LaTeX
@@ -124,5 +151,15 @@ mod tests {
         l.78         \begin{itemize}
 
         "###);
+    }
+
+    #[test]
+    fn versions() {
+        let a = Version::from_str("2.10").unwrap();
+        let b = Version::from_str("2.12.1").unwrap();
+        let c = Version::from_str("3.1.12.3").unwrap();
+        assert!(a < b);
+        assert!(b < c);
+        assert!(a < c);
     }
 }

--- a/src/pandoc/extension.rs
+++ b/src/pandoc/extension.rs
@@ -1,3 +1,5 @@
+use super::Version;
+
 #[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
 pub enum Extension {
     Strikeout,
@@ -15,6 +17,7 @@ pub enum Extension {
 }
 
 impl Extension {
+    /// The name of the extension.
     pub const fn name(&self) -> &str {
         match self {
             Extension::Strikeout => "strikeout",
@@ -28,7 +31,8 @@ impl Extension {
         }
     }
 
-    fn version_requirement(&self) -> semver::VersionReq {
+    /// Returns the pandoc version that added support for this extension.
+    fn introduced_in(&self) -> Version {
         let (major, minor, patch) = match self {
             Extension::Strikeout => (0, 10, 0),
             Extension::Footnotes => (2, 10, 1),
@@ -39,40 +43,33 @@ impl Extension {
             Extension::RawAttribute => (2, 10, 1),
             Extension::RebaseRelativePaths => (2, 14, 0),
         };
-        semver::VersionReq {
-            comparators: vec![semver::Comparator {
-                // Assumes that pandoc doesn't remove extensions once it has added them
-                op: semver::Op::GreaterEq,
-                major,
-                minor: Some(minor),
-                patch: Some(patch),
-                pre: semver::Prerelease::EMPTY,
-            }],
+        Version {
+            major,
+            minor,
+            patch,
         }
     }
 
-    pub fn check_availability(&self, pandoc: &semver::Version) -> Availability {
-        Availability::check(self.version_requirement(), pandoc)
+    pub fn check_availability(&self, pandoc: &Version) -> Availability {
+        let introduced_in = self.introduced_in();
+        if *pandoc >= introduced_in {
+            Availability::Available
+        } else {
+            Availability::Unavailable { introduced_in }
+        }
     }
 }
 
 pub enum Availability {
     Available,
-    Unavailable(semver::VersionReq),
+    Unavailable { introduced_in: Version },
 }
 
 impl Availability {
-    pub fn check(version_req: semver::VersionReq, pandoc: &semver::Version) -> Self {
-        if version_req.matches(pandoc) {
-            Availability::Available
-        } else {
-            Availability::Unavailable(version_req)
-        }
-    }
     pub fn is_available(&self) -> bool {
         match self {
             Self::Available => true,
-            Self::Unavailable(_) => false,
+            Self::Unavailable { .. } => false,
         }
     }
 }

--- a/src/pandoc/renderer.rs
+++ b/src/pandoc/renderer.rs
@@ -13,7 +13,7 @@ use tempfile::NamedTempFile;
 use crate::{
     book::Book,
     latex,
-    pandoc::{self, extension, Profile},
+    pandoc::{self, extension, Profile, Version},
     CodeConfig,
 };
 
@@ -79,10 +79,11 @@ impl Renderer {
                         format.push('+');
                         format.push_str(extension.name());
                     }
-                    extension::Availability::Unavailable(version_req) => {
+                    extension::Availability::Unavailable { introduced_in } => {
                         log::warn!(
-                            "Cannot use Pandoc extension `{}`, which may result in degraded output (requires version {}, but using {})",
-                            extension.name(), version_req, ctx.pandoc.version,
+                            "Cannot use Pandoc extension `{}`, which may result in degraded output \
+                            (introduced in version {}, but using {})",
+                            extension.name(), introduced_in, ctx.pandoc.version,
                         );
                     }
                 }
@@ -205,8 +206,12 @@ impl Renderer {
 
         let _filter_tempfile_guard: tempfile::TempPath;
         if (ctx.pandoc.enabled_extensions).contains_key(&pandoc::Extension::PipeTables) {
-            let version_req = semver::VersionReq::parse(">=2.9.2").unwrap();
-            if version_req.matches(&ctx.pandoc.version) {
+            let introduced_in = Version {
+                major: 2,
+                minor: 9,
+                patch: 2,
+            };
+            if ctx.pandoc.version >= introduced_in {
                 let mut filter = NamedTempFile::new()?;
                 write!(
                     filter,
@@ -219,7 +224,7 @@ impl Renderer {
             } else {
                 log::warn!(
                     "Cannot wrap cell contents of tables, which may result in tables overflowing the page (requires Pandoc version {}, but using {})",
-                    version_req, ctx.pandoc.version,
+                    introduced_in, ctx.pandoc.version,
                 );
             }
         }

